### PR TITLE
Unify isn() function interface for SNPs and for Genes

### LINKED ISSUE #20: `snp_interaction` uses a fixed neighbourhood for positional mapping
`snp_interaction` uses a fixed neighbourhood of 2000 nucleotides for positional mapping. Is this intentional? If so, can this constant be meaningfully named? Otherwise can the neighbourhood be a parameter?

### PR DESCRIPTION
**Pull request by MaybeJustJames**
This PR includes #14 extensions.

Closes #20

### CODE REVIEW

**Comment by GiadaLalli** [comment]
small mod: line 115

replace: A function used to replace missing values in a column either with the mode
             of the column (for discrete missing data; e.g., in SNP array data these missing                                                                      
             values are usually stored as 9 and/or -9) or its mean (for continuous data, e.g.,                                          
             expression data).

**Comment by GiadaLalli** [comment]
small mods: 

- lines [142:184]: replace "snp_info" with "unmapped_info" and "gene_info" with "mapped_info"; _let's consider to make this function internal_
- lines [190:219]: line _190_: replace def "snp_interaction" with "map_interaction"; line _191_: replace "snp_info" with "unmapped_info" and "gene_info" with "mapped_info"; line _192_ : replace "GeneInteraction" with "MappedInteraction"; lines _209:220_ : replace the new "interact_gene" with "interact_mapped", "interact_snp" with "interact_unmapped", "gene_id_1/2" with "feature 1/2"

**Comment by MaybeJustJames** [comment]
> * replace "impute" with def "mode_genotype" for consistency

Not sure what you mean here.

**Comment by GiadaLalli** [comment]
small mods: 

- line 222: replace "Metrics for SNP array" with "Metrics for unmapped discrete data"
- lines [314:365]: replace "interact_gene" with "interact_mapped", "interact_snp" with "interact_unmapped"; _this function for how I read it looks like only for SNPs, correct?_ Lioness for the dense computation is missing or it's just so modified that I cannot see it? 🤣

**Comment by GiadaLalli** [comment]
> 

we call mean_genotype the imputation for continuous data, so let's call mode_genotype the imputation for discrete data

**Comment by MaybeJustJames** [comment]
`isn()` does what both [`compute_isn()`](https://github.com/GiadaLalli/ISN-tractor/blob/gene_expression/isn_tractor/ibisn.py#L308) and [`isn_calculation_all()`](https://github.com/GiadaLalli/ISN-tractor/blob/gene_expression/isn_tractor/ibisn.py#L371) does through a single interface. I did not realise `isn_computation_all()`  was the lioness algorithm sorry. In that case they can be separated again then. Perhaps as `sparse_isn()` and `dense_isn()`.

**Comment by GiadaLalli** [comment]
here it's lioness [to be renamed isn_dense]:

def lioness(df, f):
    if not callable(f):
        raise ValueError("please use a function")
    if not isinstance(df, pd.DataFrame):
        raise ValueError("x must be a DataFrame")
    
    nrsamples = df.shape[1]
    samples = df.columns
    
    net = pearson_correlation(df.T)
    agg = net.to_numpy().flatten()
    
    # create a pandas dataframe with NaN values with dimensions of df
    lionessOutput = pd.DataFrame(np.nan, index=np.arange(df.shape[0]*df.shape[0]), 
                                 columns=["reg", "tar"]+list(samples)).astype(object)
    lionessOutput.iloc[:, 0] = np.repeat(net.columns.values, net.columns.size)
    lionessOutput.iloc[:, 1] = np.tile(net.columns.values, df.shape[0])
    #lionessOutput.iloc[:, 2:] = lionessOutput.iloc[:, 2:].apply(pd.to_numeric)
    
    for i in range(nrsamples):
        ss = pearson_correlation(pd.DataFrame(np.delete(df.T.to_numpy(), i, 0))).values.flatten() # apply Pearson on all samples minus one
        lionessOutput.iloc[:, i+2] = nrsamples*(agg-ss) + ss # apply LIONESS equation
    return lionessOutput

and this was the __compute_metric for it (yeah, from the stupid OOP version):
def __compute_metric(self, x, y=None):
            
            if self.x_type == "mapped" and self.isdense:
                if self.metric == "pearson":
                    score = x.corr(method='pearson')
                elif self.metric == "spearman":
                    score = x.corr(method='spearman')
                else:
                    raise ValueError('Wrong input for metric!')

File: isn_tractor/ibisn.py [hunk 1]
```diff
@@ -136,171 +178,220 @@ def positional_mapping(
...
-    isn = pd.DataFrame(isn, columns=[a + "_" + b for a, b in interact_gene.values])
-    return isn
+    return make_edge
+
+
+def __identity(value: _FloatLike_co) -> _FloatLike_co:
+    return value
+
+
+def isn(
```
Comment (GiadaLalli) [inline isn_tractor/ibisn.py:318]: stupid last small mod: since we called **dense_isn** the lioness-based function, let's call this one **sparse_isn**

Comment (MaybeJustJames) [reply isn_tractor/ibisn.py]: Done :)

**Pull request merged**